OpenSim/Simulation/Model/BushingForce.cpp

```diff
@@ -85,64 +79,126 @@ void BushingForce::setNull()
 {
-    constructProperty_body_1();
-    constructProperty_body_2();
-
-    //Default location and orientation (rotation sequence)
-    constructProperty_location_body_1(Vec3(0)); // body origin
-    constructProperty_orientation_body_1(Vec3(0)); // no rotation
-    constructProperty_location_body_2(Vec3(0));
-    constructProperty_orientation_body_2(Vec3(0));
+    // default bushing material properties
     constructProperty_rotational_stiffness(Vec3(0));
     constructProperty_translational_stiffness(Vec3(0));
     constructProperty_rotational_damping(Vec3(0));
     constructProperty_translational_damping(Vec3(0));
+
+    //Default frames list is empty
+    constructProperty_frames();
 }
 
 //_____________________________________________________________________________
-/**
- * Perform some set up functions that happen after the
- * object has been deserialized or copied.
- *
- * @param aModel OpenSim model containing this BushingForce.
- */
-void BushingForce::extendConnectToModel(Model& aModel)
+/*
+* Construct Structural Connectors
+*/
+void BushingForce::constructConnectors() {
+    constructConnector<PhysicalFrame>("frame1");
+    constructConnector<PhysicalFrame>("frame2");
+}
+
+void BushingForce::updateFromXMLNode( SimTK::Xml::Element& aNode,
+                                      int versionNumber)
 {
-    Super::extendConnectToModel(aModel); // base class first
-
-    // Look up the two bodies being connected by bushing by name in the
-    // model. TODO: use Connectors
-    const string& body1Name = get_body_1(); // error if unspecified
-    const string& body2Name = get_body_2();
-    _body1 = 
-        static_cast<const PhysicalFrame*>(&getModel().getComponent(body1Name));
-    _body2 =
-        static_cast<const PhysicalFrame*>(&getModel().getComponent(body2Name));
+    int documentVersion = versionNumber;
+    bool converting = false;
+    if (documentVersion < XMLDocument::getLatestVersion()){
+        if (documentVersion < 30503){
+            // replace old properties with latest use of PhysicalOffsetFrames properties
+            SimTK::Xml::element_iterator body1Element =
+                aNode.element_begin("body_1");
+            SimTK::Xml::element_iterator body2Element =
+                aNode.element_begin("body_2");
+            SimTK::Xml::element_iterator locBody1Elt =
+                aNode.element_begin("location_body_1");
+            SimTK::Xml::element_iterator orientBody1Elt =
+                aNode.element_begin("orientation_body_1");
+            SimTK::Xml::element_iterator locBody2Elt =
+                aNode.element_begin("location_body_2");
+            SimTK::Xml::element_iterator orientBody2Elt =
+                aNode.element_begin("orientation_body_2");
+
+            //Create two new PhysicalOffsetFrames
+            PhysicalOffsetFrame frame1;
+            PhysicalOffsetFrame frame2;
+            frame1.setName("frame1_offset");
+            frame2.setName("frame2_offset");
```

Comment:
Would it make sense to *not* create these frames if the pose was (0,0,0,0,0,0) or not present? Could use the body frames directly in that case rather than locally-owned ones.